src/Microsoft.OData.Core/ODataMetadataJsonOutputContext.cs

```diff
@@ -156,19 +180,55 @@ protected override void Dispose(bool disposing)
             {
                 if (this.jsonWriter != null)
                 {
-                    this.jsonWriter.Flush();
-                    this.jsonWriter.Dispose();
+                    if (this.asynchronousOutputStream != null)
+                    {
+                        this.asynchronousOutputStream.FlushSync();
+                        this.asynchronousOutputStream.Dispose();
+
+                        this.jsonWriter.FlushAsync();
+                        this.jsonWriter.DisposeAsync();

```

Comment:
We're not waiting for the async operations to complete? Or is it okay to just fire and forget since the asynchronous stream has already been flushed and disposed?

Reply:
Below is the code for DisposeAysnc():

```C#
        public async ValueTask DisposeAsync()
        {
            if (_stream == null)
            {
                // The conditions are ordered with stream first as that would be the most common mode
                if (_output == null)
                {
                    return;
                }
            }

            await FlushAsync().ConfigureAwait(false);
            ResetHelper();

            _stream = null;
            _arrayBufferWriter = null;
            _output = null;
        }
```

In this case, i think we don't need to call await, and we can NOT call await in a non-async method. @mikepizzo

---
In reply to: [556555576](https://github.com/OData/odata.net/pull/1966#discussion_r556555576) [](ancestors = 556555576)